Bluetooth: llcp: Fix PHY UPD notification send before instant on is air

ppryga-nordic opened the pull request:
PHY update control procedure has to send a notification towards
Host when the PHY is actually changed, when the instant happens.
Control procedures are handled in a prepare phase of a connection
event, hence notifications are send towards host to early.

The fix for that is to postpone send of a notification.
New state has been introduced to remote requested PHY update
control procedure: RP_PU_STATE_WAIT_INSTANT_ON_AIR.
Also new event has been introduced: RP_PU_EVT_NTF, that is used
to inform the PHY update state machine that notification may
be send.

There are two places where the notification events may be generated
ull_conn_rx, just before first received PDU is send towards Host
or ull_conn_done, in case there were no PDU received.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: llcp: Fix PHY UPD ntf send before instant on air

PHY update control procedure has to send a notification towards
Host when the PHY is actually changed, when the instant happens.
Control procedures are handled in a prepare phase of a connection
event, hence notifications are send towards host to early.

The fix for that is to postpone send of a notification.
New state has been introduced to remote requested PHY update
control procedure: RP_PU_STATE_WAIT_INSTANT_ON_AIR.
Also new event has been introduced: RP_PU_EVT_NTF, that is used
to inform the PHY update state machine that notification may
be send.

There are two places where the notification events may be generated
ull_conn_rx, just before first received PDU is send towards Host
or ull_conn_done, in case there were no PDU received.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

thoh-ot commented:
Looks very cool

ppryga commented:
Completed implementation of delay of Host notification for remote PHY update request. Added implementation for delay of a notification for local PHY update request.

ppryga-nordic committed:
tests: Bluetooth: llpc: Add instant on air notification for PHY UPD

Unit tests does not maintain connection events hence there is no
notification to PHY upodate state machine about instant on air
event. That has to be simulated by explicit call to ull_cp_tx_nrf.

In unit tests the event_done function is responsible for handling
all steps done at end of connection event, so it is the best place
to simulate the instant on air event and send it to LLCP state machine.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: llcp: Add missing at-instant conn event processing

The test_phy_update_central_rem_collision didn't have processing
of an "at-instant" connection event. Thas causes the remote requesed
PHY update to hand in waiting for instant on air state and whole
test to fail.

To fix the problem, proper processing of "at-instant" connection evnet
has to be added. In consequence the local PHY update procedure, that
as waiting for end of the remote procedure is delayed by one connection
event.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Fixed failing unit tests.

ppryga-nordic committed:
tests: Bluetooth: llcp: Add handling delayed PHY UPD compl in CTE REQ UT

The PHY Update control procedure completion is delayed to event
"instant-on-air". That has to be handled in CTE Request control
procedure unit test.

The changes required are:
- Added handling of event done for "at-instant" connection event.
  This is required to call ull_cp_tx_ntf function to send an
  envet "instant-on-air" to PHY Update procedure state machine.
  This is the delayed completion of PHY Update procedure.
- Remote requested PHY Update completion is also delayed and completed
  on "instant-on-air" event, hence the local CTE Request is not executed
  at the same connection event as PHY Update completion. This caused
  changes to check_phy_update_and_cte_req_complete function. The functin
  does not hanle CTE Request completion anymore. This also implies
  changes to run_phy_update_central and run_phy_update_peripheral.
  These function do not need to get CTE request argument anymore
  because it was just passed to check_phy_update_and_cte_req_complete.
  The check_phy_update_cte_req_complete was renamed to check_phy_update.
  The test test_cte_req_wait_for_remote_phy_update_complete was changed
  to execute and verify local CTE request procedure explicitly after
  completion of PHY Update.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
I've fixed CTE Request unit tests that are responsible for verification of mutual blocking with PHY Update procedure, hence they were affected by the PR.

ppryga-nordic committed:
tests: Bluetooth: llpc: Add at-instant conn evt proc in collision UT

Unit tests reposnsible for verification of collisions mitigation
didn't have handling for delayed completion of PHY Update control
procedure. The PHY Update is completed on event generated when
"at-instant" connection event is on air.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: llpc: Fix enc proc assert on released ctx num

Due to added call to ull_cp_tx_nrf call in common function event_done,
encryption test test_encryption_start_periph_rem_no_ltk failed with
assert caused by unexpected number of free control procedure contexts.

That is caused by internal call to rr_check_done and lr_check_done
in the ull_cp_tx_nrf that frees last unreleased context.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Fixed `ctrl_encryption` unit test.
